pkg/kubelet/logs/container_log_manager.go

```diff
@@ -0,0 +1,401 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logs
+
+import (
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/wait"
+	internalapi "k8s.io/kubernetes/pkg/kubelet/apis/cri"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/runtime/v1alpha2"
+)
+
+const (
+	// logMonitorPeriod is the period container log manager monitors
+	// container logs and performs log rotation.
+	logMonitorPeriod = 10 * time.Second
+	// timestampFormat is format of the timestamp suffix for rotated log.
+	// See https://golang.org/pkg/time/#Time.Format.
+	timestampFormat = "20060102-150405"
+	// compressSuffix is the suffix for compressed log.
+	compressSuffix = ".gz"
+	// tmpSuffix is the suffix for temporary file.
+	tmpSuffix = ".tmp"
+)
+
+// ContainerLogManager manages lifecycle of all container logs.
+//
+// Implementation is thread-safe.
+type ContainerLogManager interface {
+	// TODO(random-liu): Add RotateLogs function and call it under disk pressure.
+	// Start container log manager.
+	Start()
+}
+
+// LogRotatePolicy is a policy for container log rotation. The policy applies to all
+// containers managed by kubelet.
+type LogRotatePolicy struct {
+	// MaxSize in bytes of the container log file before it is rotated. Negative
+	// number means to disable container log rotation.
+	MaxSize int64
+	// MaxFiles is the maximum number of log files that can be present.
+	// If rotating the logs creates excess files, the oldest file is removed.
+	MaxFiles int
+}
+
+// GetAllLogs gets all inuse (rotated/compressed) logs for a specific container log.
+// Returned logs are sorted in oldest to newest order.
+// TODO(#59902): Leverage this function to support log rotation in `kubectl logs`.
+func GetAllLogs(log string) ([]string, error) {
+	// pattern is used to match all rotated files.
+	pattern := fmt.Sprintf("%s.*", log)
+	logs, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all log files with pattern %q: %v", pattern, err)
+	}
+	inuse, _ := filterUnusedLogs(logs)
+	sort.Strings(inuse)
+	return append(inuse, log), nil
+}
+
+// compressReadCloser wraps gzip.Reader with a function to close file handler.
+type compressReadCloser struct {
+	f *os.File
+	*gzip.Reader
+}
+
+func (rc *compressReadCloser) Close() error {
+	ferr := rc.f.Close()
+	rerr := rc.Reader.Close()
+	if ferr != nil {
+		return ferr
+	}
+	if rerr != nil {
+		return rerr
+	}
+	return nil
+}
+
+// UncompressLog compresses a compressed log and return a readcloser for the
+// stream of the uncompressed content.
+// TODO(#59902): Leverage this function to support log rotation in `kubectl logs`.
+func UncompressLog(log string) (_ io.ReadCloser, err error) {
```

Comment:
you dont use the err defined in the function declaration, so remove it.

Reply:
It is used in `defer`.

Reply:
ahh, neat.  Will `err != nil` below (line 113) return true when we return an error on line 109?

Reply:
Yeah. Let me rename this var to `retErr`, which makes the code less confusing.

Reply:
Done